### PR TITLE
Scoreboard Use Objectives and Animation Command

### PR DESCRIPTION
**Pull request by Melodi17**
Scoreboard objectives for using blocks and items, and a command for playing animations on entities.

### CODE REVIEW

**Comment by TheDeathlyCow** [comment]
I think there's a chance that the swing command is somewhat redundant as [Carpet](https://github.com/gnembon/fabric-carpet) already has something similar under its `/player` command.

**Comment by TheDeathlyCow** [comment]
Also, I renamed the left click and right click stats to attack and use respectively, as they can be rebound to different keys. The server does not know what the difference between them is.

**Comment by TheDeathlyCow** [comment]
I've been having a look at how Minecraft handles click detection and server/client syncing. It seems like it could be rather difficult to detect key bind events like right, left, and middle clicking on the server without needing to create a new packet (which would mean the mod would require clients to install it). This is a problem because it can lead to some clicks not being detected at all (e.g., left clicking air), or some events being counted twice (e.g., right-clicking on a block with an item in hand). 

Also, the use block event seems to fire twice when interacting with a block that doesn't have a use (e.g. dirt). This seems to happen independently of the double-counted item use on a block, as holding an item and clicking dirt increments the stat by 3.

**Comment by Melodi17** [comment]
For the swing command, I think we should keep it in as it can be useful if the user does not have any other mods and it doesn't have much of a performance impact. The renaming of the left and right click events to use and attack makes sense, thanks for that. For the pick block, I can't see much use for being able to detect it anyway and detecting it without client-side modifications seems to be not possible in some scenarios and relies on specific circumstances (e.g. the player being in survival). For the multiple event fires issue, maybe we should add another stat for each individual event (UseItem, UseBlock, UseEntity, AttackEntity, AttackBLock) and possibly keep the central Use and Attack stats.

**Comment by TheDeathlyCow** [comment]
I think it would be better to just have the individual stats, and drop the central ones. If you use an item on a block, it should only count towards a central use stat once - but it doesn't. This would be really weird and confusing in my opinion. Also, it still doesn't track things like clicking air, which is still 'attacking'. And you'd still easily be able to tell when an item is used, just check if any of the stat-tracked scores are greater than 1 using a `minecraft:alternative` predicate for example. 

Also, somewhat related, a `tdcdata:use_item` trigger similar to `minecraft:using_item` but for any item could also be really useful here.

**Comment by Melodi17** [comment]
For the trigger/advancement I tried to but I couldn't understand how to implement it, sorry.

**Comment by TheDeathlyCow** [comment]
Thanks! Don't worry, I can handle the advancement.

**Comment by TheDeathlyCow** [comment]
I think `/swing` could be expanded upon further as a general animation command feature similar to Bedrock's [/playanimation](https://minecraft.fandom.com/wiki/Commands/playanimation) command, where other animations could be played like jumping or dying.

**Comment by Melodi17** [comment]
I really like that idea. I believe it could be really useful.

**Comment by TheDeathlyCow** [comment]
Here are my thoughts on `/var`... 

I think the idea of having a way to do string manipulation is cool. However, I think its current implementation is a bit messy and full of redundancies. A better implementation (which should be in a new PR from this, this PR is already getting way too many things at once) would make use of NBT data. NBT data can store a lot of things already - including strings. Operating on these strings, I think, is a much better thing to do than trying to implement a whole new system like this ourselves. Datapack developers are already familiar with dealing with NBT data, and it has a lot of hooks into things like entities, block entities, and storage that would be useful. If we want to do other things like getting block IDs into NBT, then that can be handled separately from this system. 

Execute if var should then instead be transformed into comparison for NBT objects, as there isn't really a way of doing this in vanilla as far as I know. This might be a bit painful to implement as NBT data can be a bit annoying to handle through Java code, but its certainly worth trying. 

Also the `VarCommand` class is a bit of a god class. It's trying to handle vars *and* the var command. These things should be separate, a `Var` class should be able to handle the parsing and interpreting of var data, and the `VarCommand` class should only be a facade for calling those methods in-game. 

To sum up: The var idea is cool but needs some work. Its current implementation is rather messy and hard to understand. It would be better implemented through the NBT system, as that is more familiar to data pack developers and works better with the existing codebase.

**Review by TheDeathlyCow — APPROVED** [review]
